Subject: nfs: fix missing preemption check
From: Thomas Gleixner <tglx@linutronix.de>
Date: Sun, 27 Jul 2008 00:54:19 +0200

NFS iostats use get_cpu()/put_cpu_no_preempt(). That misses a
preemption check for no good reason and introduces long latencies when
a wakeup of a higher priority task happens in the preempt disabled
region.

Signed-off-by: Thomas Gleixner <tglx@linutronix.de>
Signed-off-by: Ingo Molnar <mingo@elte.hu>
---
 fs/nfs/iostat.h |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

Index: linux-2.6-tip/fs/nfs/iostat.h
===================================================================
--- linux-2.6-tip.orig/fs/nfs/iostat.h
+++ linux-2.6-tip/fs/nfs/iostat.h
@@ -28,7 +28,7 @@ static inline void nfs_inc_server_stats(
 	cpu = get_cpu();
 	iostats = per_cpu_ptr(server->io_stats, cpu);
 	iostats->events[stat]++;
-	put_cpu_no_resched();
+	put_cpu();
 }
 
 static inline void nfs_inc_stats(const struct inode *inode,
@@ -47,7 +47,7 @@ static inline void nfs_add_server_stats(
 	cpu = get_cpu();
 	iostats = per_cpu_ptr(server->io_stats, cpu);
 	iostats->bytes[stat] += addend;
-	put_cpu_no_resched();
+	put_cpu();
 }
 
 static inline void nfs_add_stats(const struct inode *inode,
